src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketOperationResult.cs

```diff
@@ -0,0 +1,28 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Net.Sockets;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
+{
+    internal readonly struct SocketOperationResult
+    {
+        public readonly SocketException? SocketError;
+
+        public readonly int BytesTransferred;
+
+        public readonly bool HasError => SocketError != null;
```

Comment:
Use `[MemberNotNullWhen(true, nameof(SocketError))]` which should get rid of all the `!` usage.

Reply:
Took me some time